### ISSUE TITLE
run.sh failure due to yarn

### ISSUE THREAD

**Issue by DanBerrios** [galaxyproject/galaxy#10149]
We are seeing run.sh fail with these yarn errors -- any idea what's going on?:

```
    ERROR in ./galaxy/style/scss/base.scss (./node_modules/css-loader/dist/cjs.js??ref--8-1!./node_modules/postcss-loader/src??ref--8-2!./node_      modules/sass-loader/lib/loader.js??ref--8-3!./galaxy/style/scss/base.scss)
    Module build failed (from ./node_modules/sass-loader/lib/loader.js):

    @import "../../../node_modules/@fortawesome/fontawesome-free/scss/fontawesome";
    ^
          File to import not found or unreadable: ../../../node_modules/@fortawesome/fontawesome-free/scss/fontawesome.
          in /data/galaxy/client/galaxy/style/scss/base.scss (line 15, column 1)
Child mini-css-extract-plugin node_modules/css-loader/dist/cjs.js??ref--8-1!node_modules/vue-loader/lib/loaders/stylePostLoader.js!node_modules      /postcss-loader/src/index.js??ref--8-2!node_modules/sass-loader/lib/loader.js??ref--8-3!node_modules/vue-loader/lib/index.js??vue-loader-option      s!galaxy/scripts/components/Markdown/Markdown.vue?vue&type=style&index=0&lang=scss&:
    Entrypoint mini-css-extract-plugin = *


...
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

**Comment by dannon** [comment]
@DanBerrios I haven't seen this one before, but we can debug this a little.  Before that, do you know what changed since last it worked?  Did you try to update Galaxy, install any local patches, etc?

In any event, first I'd try a standalone client build, doing something like the following:
```
cd <galaxy_root>
source .venv/bin/activate
cd client
yarn install --check-files
yarn run build
```

This activates your virtualenv, and installs client dependencies, and then builds the Galaxy client.

**Comment by DanBerrios** [comment]
Thanks for helping.  We have been tweaking the client styling and somehow we got in a state where there's mismatches in some of the styling files and what's installed by node I think.  We rolled back the styling changes in base.scss, and I had already done a yarn update, and that cleared some of the errors...but we are still left with some client build errors.  Here's one of them....can you point us in the direction of how to fix this?

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/command-aliases' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 6:9-49
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

**Comment by dannon** [comment]
Without knowing the full extent of what local tweaks and commands have been run, I'd suggest just removing node_modules and rebuilding.  
```
source .venv/bin/activate
cd client
rm -rf node_modules
yarn install --check-files
yarn run build
```

**Comment by DanBerrios** [comment]
OK, I did that.... the same set of errors still persist after yarn run build is executed:

```
ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/command-aliases' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 6:9-49
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/convert' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 7:9-41
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/latex/commands/1args/index' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 14:9-60
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/options' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 8:9-41
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/tex/KnownCommand' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 11:9-50
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/tex/KnownCommand0Args' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 12:9-55
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/tex/KnownCommand1Args' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 13:9-55
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/unknown-command' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 9:9-49
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

ERROR in ./node_modules/latex-to-unicode-converter/index.js
Module not found: Error: Can't resolve './ts-compiled/util' in '/data/galaxy/client/node_modules/latex-to-unicode-converter'
 @ ./node_modules/latex-to-unicode-converter/index.js 10:9-38
 @ ./node_modules/babel-loader/lib??ref--1!./node_modules/vue-loader/lib??vue-loader-options!./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue?vue&type=script&lang=js&
 @ ./galaxy/scripts/components/Citations.vue
 @ ./galaxy/scripts/entry/analysis/AnalysisRouter.js
 @ ./galaxy/scripts/entry/analysis/index.js
 @ multi polyfills bundleEntries entry/analysis

...
Child mini-css-extract-plugin node_modules/css-loader/dist/cjs.js??ref--8-1!node_modules/vue-loader/lib/loaders/stylePostLoader.js!node_modules/postcss-loader/src/index.js??ref--8-2!node_modules/sass-loader/dist/cjs.js??ref--8-3!node_modules/vue-loader/lib/index.js??vue-loader-options!galaxy/scripts/components/User/CloudAuth/CloudAuth.vue?vue&type=style&index=0&lang=scss&:
    Entrypoint mini-css-extract-plugin = *
       9 modules
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by dannon** [comment]
What exact revision of Galaxy is this again, and if there are still outstanding local changes is there any way I can take a look at those?

The latest `dev` branch (which I don't recommend for a production instance) doesn't use `latex-to-unicode-converter` anymore, and I'm wondering if that's where some of the mixup is here.  When I check out `release_20.05` and reinstall clean node_modules (via that procedure above), I do see the files available in `/galaxy/client/node_modules/latex-to-unicode-converter/ts-compiled/*`

**Comment by DanBerrios** [comment]
We have 20.01.

**Comment by DanBerrios** [comment]
I'm happy to do a screen share with you to debug...just let me know what time is good for you and what email to send invite to.

**Comment by DanBerrios** [comment]
After doing your yarn rebuild procedure above, the file ./ts-compiled/command-aliases still isn't on our server....  But we have 20.01 not 20.05

**Comment by dannon** [comment]
I'll have to check out 20.01 in a few;  can you check the precise version specified in the yarn lock file for this dependency, and what's actually in the directory expected?  Mine looks like this:
```
[yoplait@incus:~/work/galaxy/client] cat yarn.lock | grep latex-to-unicode     130 ↵release_20.05 ✭
latex-to-unicode-converter@^0.5.2:
  resolved "https://registry.yarnpkg.com/latex-to-unicode-converter/-/latex-to-unicode-converter-0.5.2.tgz#6c3702f66ab7362fb6ec3dcfeb8bbbf28cf3b945"
[yoplait@incus:~/work/galaxy/client] ls node_modules/latex-to-unicode-converter/ts-compiled
command-aliases.d.ts html                 options.d.ts         unknown-command.d.ts
command-aliases.js   index.d.ts           options.js           unknown-command.js
convert.d.ts         index.js             tex                  util.d.ts
convert.js           latex                unicode              util.js
```

**Comment by DanBerrios** [comment]
```
(.venv) -bash-4.2$ cat yarn.lock | grep latex-to-unicode
latex-to-unicode-converter@^0.5.2:
  resolved "https://registry.yarnpkg.com/latex-to-unicode-converter/-/latex-to-unicode-converter-0.5.4.tgz#b9f89e9cf69c19d2addb8929cd663fa4f3a0846c"
(.venv) -bash-4.2$ ls node_modules/latex-to-unicode-converter/ts-compiled
index.d.ts  index.js
(.venv) -bash-4.2$
```
Perplexing!

**Comment by DanBerrios** [comment]
And...no errors in yarn install step....

**Comment by DanBerrios** [comment]
yarn build has various warnings..... but no errors....

**Comment by dannon** [comment]
Ahh!  It's actually installing 0.5.4 instead of 0.5.2.  

Can you try this change?   https://gist.github.com/90dc6861f5e04ea5e9c8cfc578cf277f

One reason we swapped away from this package was the deprecation, and it looks like the later versions have dropped all these other files that were being relied on, with the final one simply not working at all.

**Comment by dannon** [comment]
If this works, the fix in https://github.com/galaxyproject/galaxy/pull/10157 *should* address it and will get the patch into the main 20.01 branch.

**Comment by DanBerrios** [comment]
It DID fix our build issues, thanks so much.  We are back in business.  I forget..do the admin docs discuss how to upgrade Galaxy? If so, we should probably upgrade to latest, right?

**Comment by dannon** [comment]
Hooray!  I'd recommend following the latest stable release, yes.  The update instructions *should* be completely painless from 20.01 -> 20.05:  https://docs.galaxyproject.org/en/master/releases/20.05_announce.html#get-galaxy

Slightly more in-depth instructions are at https://galaxyproject.org/admin/maintenance/, or via slides at https://training.galaxyproject.org/training-material/topics/admin/tutorials/upgrading/slides.html#1

**Comment by DanBerrios** [comment]
Thanks!  We will be doing that...

**Comment by DanBerrios** [comment]
Feel free to close from our end (or leave open if you are waiting on pull request to go through)
